File: products/conversations/backend/plan_tags.py
# The support team's plan/priority tiering, derived from ticket tags. Group
# order IS the priority order (index 0 = Triage = highest). A ticket takes the
# highest-priority group with a matching tag; tickets with no matching tag
# rank with Triage (they still need routing).
#
# MUST stay in lockstep with the frontend copy in
# products/conversations/frontend/scenes/tickets/planTags.ts (the column's
# labels live there; ranks here). Tag vocabulary sources: the "Customer and
# account tags" section of https://posthog.com/handbook/support/posthog-support,
# the live SLA workflow's tags, and the Zendesk import's tag names.
from django.db.models import Case, Exists, IntegerField, OuterRef, Value, When

# rank → exact tag names (no prefix matching)
PLAN_GROUP_TAGS: list[list[str]] = [
    ["support_needs_triage"],  # 0 Triage (also the untagged fallback)
    ["churn_risk"],  # 1 Churn risk
    ["plan_top20", "top_20"],  # 2 Top 20
    [  # 3 Enterprise
        "plan_enterprise",
        "goodwill_enterprise",
        "unknown_slack_default_enterprise",
        "unknown_msteams_default_enterprise",
    ],
    ["plan_onboarding", "new_customer_onboarding"],  # 4 Onboarding
    ["plan_scale", "plan_teams_legacy", "plan_teams", "plan_yc"],  # 5 Scale & Teams & YC
    [  # 6 Boost & Startup & Pay-as-you-go paying
        "plan_boost",
        "plan_startup",
        "plan_pay-as-you-go_paying",
        "plan_pay-as-you-go",
        "plan_paid",
    ],
    ["plan_pay-as-you-go_free"],  # 7 Pay-as-you-go free
    ["plan_free"],  # 8 Free plan
    ["community"],  # 9 Community
]


def plan_rank_annotation() -> Case:
    """A per-ticket plan rank for ORDER BY: the first (highest-priority) group
    with a matching tag wins, courtesy of Case evaluating Whens in order.
    Untagged/unmatched tickets take the default rank 0 (Triage).

    Perf note: this is one correlated EXISTS per group (10 today). Each can be
    served by TaggedItem's partial unique index on (tag, ticket) probed
    tag-first (the group's tag ids are few), but there is no ticket-leading
    index — if staff usage makes this sort hot at scale, consider adding one.
    """
    from posthog.models.tagged_item import TaggedItem

    return Case(
        *[
            When(
                Exists(TaggedItem.objects.filter(ticket=OuterRef("pk"), tag__name__in=tags)),
                then=Value(rank),
            )
            for rank, tags in enumerate(PLAN_GROUP_TAGS)
        ],
        default=Value(0),
        output_field=IntegerField(),
    )

File: products/conversations/frontend/scenes/tickets/planTags.ts
// The customer's plan tier, derived from ticket tags — the support team's
// triage ladder. Group order IS the priority order (Triage first, Community
// last): a ticket takes the highest-priority group with a matching tag, and
// tickets with no matching tag fall into Triage (they still need routing).
//
// Tag vocabulary sources, all kept so every era of ticket groups correctly:
// - The "Customer and account tags" section of
//   https://posthog.com/handbook/support/posthog-support (source of truth)
// - The live "Support:: SLA:: identify plan types and set NRT SLAs" workflow
//   (top_20, plan_teams, plan_paid, new_customer_onboarding, ...)
// - The Zendesk import (plan_top20, plan_teams_legacy, plan_pay-as-you-go*,
//   ... — surveyed against the zendesk.tickets warehouse table, 2026-07-20)
//
// Matching is exact (no prefixes): plan_enterprise_trial is NOT Enterprise.

export interface PlanGroup {
    label: string
    tags: string[]
}

export const PLAN_GROUPS: PlanGroup[] = [
    { label: 'Triage', tags: ['support_needs_triage'] },
    { label: 'Churn risk', tags: ['churn_risk'] },
    { label: 'Top 20', tags: ['plan_top20', 'top_20'] },
    {
        label: 'Enterprise',
        tags: [
            'plan_enterprise',
            'goodwill_enterprise',
            'unknown_slack_default_enterprise',
            'unknown_msteams_default_enterprise',
        ],
    },
    { label: 'Onboarding', tags: ['plan_onboarding', 'new_customer_onboarding'] },
    { label: 'Scale & Teams & YC', tags: ['plan_scale', 'plan_teams_legacy', 'plan_teams', 'plan_yc'] },
    {
        label: 'Boost & Startup & Pay-as-you-go paying',
        tags: ['plan_boost', 'plan_startup', 'plan_pay-as-you-go_paying', 'plan_pay-as-you-go', 'plan_paid'],
    },
    { label: 'Pay-as-you-go free', tags: ['plan_pay-as-you-go_free'] },
    { label: 'Free plan', tags: ['plan_free'] },
    { label: 'Community', tags: ['community'] },
]

const TAG_RANK = new Map<string, number>(PLAN_GROUPS.flatMap((group, rank) => group.tags.map((tag) => [tag, rank])))

/** Index into PLAN_GROUPS of the ticket's group — the best (lowest) rank
 *  across its tags, or Triage (0) when none match. */
export function planRank(tags: string[] | undefined): number {
    let best = Number.POSITIVE_INFINITY
    for (const tag of tags ?? []) {
        const rank = TAG_RANK.get(tag)
        if (rank !== undefined && rank < best) {
            best = rank
        }
    }
    return best === Number.POSITIVE_INFINITY ? 0 : best
}

export function planLabel(tags: string[] | undefined): string {
    return PLAN_GROUPS[planRank(tags)].label
}
